[BUG] test_sample_produce_empty_batch failed in dataproc

**Describe the bug**
new test was introduced in #3789

This failure showed only in dataproc, and was not seen from other ENVs yet
```
cpu = [Row(a=None), Row(a='\x82âæb÷(´û'), Row(a='Ôq=`È§\x05:'), Row(a='ãç\x05\x82Â\x08L\xad')]
gpu = [Row(a='\x82âæb÷(´û'), Row(a='Ôq=`È§\x05:'), Row(a='ãç\x05\x82Â\x08L\xad')]
```

assert error,
```
[2021-10-19T11:21:18.998Z] =================================== FAILURES ===================================
[2021-10-19T11:21:18.998Z] ___________________ test_sample_produce_empty_batch[String] ____________________
[2021-10-19T11:21:18.998Z] 
[2021-10-19T11:21:18.998Z] data_gen = String
[2021-10-19T11:21:18.998Z] 
[2021-10-19T11:21:18.998Z]     @ignore_order
[2021-10-19T11:21:18.998Z]     @pytest.mark.parametrize('data_gen', [string_gen], ids=idfn)
[2021-10-19T11:21:18.998Z]     def test_sample_produce_empty_batch(data_gen):
[2021-10-19T11:21:18.998Z] >       assert_gpu_and_cpu_are_equal_collect(
[2021-10-19T11:21:18.998Z]             # length = 4 will generate empty batch after sample
[2021-10-19T11:21:18.998Z]             lambda spark: unary_op_df(spark, data_gen, length= 4).sample(fraction = 0.9, seed = 1)
[2021-10-19T11:21:18.998Z]         )
[2021-10-19T11:21:18.998Z] 
[2021-10-19T11:21:18.998Z] integration_tests/src/main/python/sample_test.py:29: 
[2021-10-19T11:21:18.998Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[2021-10-19T11:21:18.998Z] integration_tests/src/main/python/asserts.py:501: in assert_gpu_and_cpu_are_equal_collect
[2021-10-19T11:21:18.998Z]     _assert_gpu_and_cpu_are_equal(func, 'COLLECT', conf=conf, is_cpu_first=is_cpu_first)
[2021-10-19T11:21:18.998Z] integration_tests/src/main/python/asserts.py:432: in _assert_gpu_and_cpu_are_equal
[2021-10-19T11:21:18.998Z]     assert_equal(from_cpu, from_gpu)
[2021-10-19T11:21:18.998Z] integration_tests/src/main/python/asserts.py:101: in assert_equal
[2021-10-19T11:21:18.998Z]     _assert_equal(cpu, gpu, float_check=get_float_check(), path=[])
[2021-10-19T11:21:18.998Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[2021-10-19T11:21:18.998Z] 
[2021-10-19T11:21:18.998Z] cpu = [Row(a=None), Row(a='\x82âæb÷(´û'), Row(a='Ôq=`È§\x05:'), Row(a='ãç\x05\x82Â\x08L\xad')]
[2021-10-19T11:21:18.998Z] gpu = [Row(a='\x82âæb÷(´û'), Row(a='Ôq=`È§\x05:'), Row(a='ãç\x05\x82Â\x08L\xad')]
[2021-10-19T11:21:18.998Z] float_check = <function get_float_check.<locals>.<lambda> at 0x7f12219a4040>
[2021-10-19T11:21:18.998Z] path = []
[2021-10-19T11:21:18.998Z] 
[2021-10-19T11:21:18.998Z]     def _assert_equal(cpu, gpu, float_check, path):
[2021-10-19T11:21:18.998Z]         t = type(cpu)
[2021-10-19T11:21:18.998Z]         if (t is Row):
[2021-10-19T11:21:18.998Z]             assert len(cpu) == len(gpu), "CPU and GPU row have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))
[2021-10-19T11:21:18.998Z]             if hasattr(cpu, "__fields__") and hasattr(gpu, "__fields__"):
[2021-10-19T11:21:18.998Z]                 assert cpu.__fields__ == gpu.__fields__, "CPU and GPU row have different fields at {} CPU: {} GPU: {}".format(path, cpu.__fields__, gpu.__fields__)
[2021-10-19T11:21:18.998Z]                 for field in cpu.__fields__:
[2021-10-19T11:21:18.998Z]                     _assert_equal(cpu[field], gpu[field], float_check, path + [field])
[2021-10-19T11:21:18.998Z]             else:
[2021-10-19T11:21:18.998Z]                 for index in range(len(cpu)):
[2021-10-19T11:21:18.998Z]                     _assert_equal(cpu[index], gpu[index], float_check, path + [index])
[2021-10-19T11:21:18.998Z]         elif (t is list):
[2021-10-19T11:21:18.998Z] >           assert len(cpu) == len(gpu), "CPU and GPU list have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))
[2021-10-19T11:21:18.998Z] E           AssertionError: CPU and GPU list have different lengths at [] CPU: 4 GPU: 3
```

--- a/integration_tests/src/main/python/sample_test.py
+++ b/integration_tests/src/main/python/sample_test.py
@@ -25,28 +25,29 @@ basic_gens = all_gen + [NullGen()]
 # If sample exec can't handle empty batch, will trigger "Input table cannot be empty" error
 @ignore_order
 @pytest.mark.parametrize('data_gen', [string_gen], ids=idfn)
-@pytest.mark.xfail(reason='https://github.com/NVIDIA/spark-rapids/issues/3892')
 def test_sample_produce_empty_batch(data_gen):
     assert_gpu_and_cpu_are_equal_collect(
-        # length = 4 will generate empty batch after sample
-        lambda spark: unary_op_df(spark, data_gen, length= 4).sample(fraction = 0.9, seed = 1)
+        # length = 4 will generate empty batch after sample;
+        # set num_slices as constant,
+        # num_slices is not same in different mode, such as local, yarn, Mesos
+        lambda spark: unary_op_df(spark, data_gen, length= 4, num_slices = 10)
+            .sample(fraction = 0.9, seed = 1)
     )
 
 # the following cases is the normal cases and do not use @ignore_order
 nested_gens = array_gens_sample + struct_gens_sample + map_gens_sample
 @pytest.mark.parametrize('data_gen', basic_gens + nested_gens, ids=idfn)
-@pytest.mark.xfail(reason='https://github.com/NVIDIA/spark-rapids/issues/3892')
 def test_sample(data_gen):
     assert_gpu_and_cpu_are_equal_collect(
-        lambda spark: unary_op_df(spark, data_gen).sample(fraction = 0.9, seed = 1),
+        lambda spark: unary_op_df(spark, data_gen, num_slices = 10)
+            .sample(fraction = 0.9, seed = 1),
         conf={'spark.sql.legacy.allowNegativeScaleOfDecimal': True}
     )
 
 @pytest.mark.parametrize('data_gen', basic_gens + nested_gens, ids=idfn)
-@pytest.mark.xfail(reason='https://github.com/NVIDIA/spark-rapids/issues/3892')
 def test_sample_with_replacement(data_gen):
     assert_gpu_and_cpu_are_equal_collect(
-        lambda spark: unary_op_df(spark, data_gen).sample(
+        lambda spark: unary_op_df(spark, data_gen, num_slices = 10).sample(
             withReplacement =True, fraction = 0.5, seed = 1),
         conf={'spark.sql.legacy.allowNegativeScaleOfDecimal': True}
     )

